### PR TITLE
Add support for a01nyub

### PR DESCRIPTION
**Pull request by MrSuicideParrot**
# What does this implement/fix?

This pull request adds support for the [a01nyub waterproof ultrasonic distance sensor](https://wiki.dfrobot.com/A01NYUB%20Waterproof%20Ultrasonic%20Sensor%20SKU:%20SEN0313) by DFRobot .

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2934

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
uart:
  rx_pin: 19
  baud_rate: 9600

sensor:
  - platform: a01nyub
    name: "Distance"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by raccoonia** [comment]
I have tested the support of this sensor for a few weeks now, it works great!
The yamlcode I am using on a ESP32 for testing is the following:

```
external_components:
  - source: github://pr#4863
    components: [ a01nyub ]
uart:
    rx_pin: 25
    baud_rate: 9600
sensor:
  - platform: a01nyub
    id: a01nyub_sensor
    name: "Distance"
    unit_of_measurement: 'm'
    accuracy_decimals: 2
    filters:
      - sliding_window_moving_average:
          window_size: 20 
          send_every: 20 
          send_first_at: 1 
```

File: esphome/components/a01nyub/a01nyub.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  // Nothing really public.
+
+  // ========== INTERNAL METHODS ==========
+  void loop() override;
+  void dump_config() override;
+
+ protected:
+  void check_buffer_();
+
+  std::string buffer_;
```
Comment (jesserockz) [inline esphome/components/a01nyub/a01nyub.h]: You don't need the string functionality here
```suggestion
  std::vector<uint8_t> buffer_;
```

File: esphome/components/a01nyub/a01nyub.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+namespace a01nyub {
+
+static const char *const TAG = "a01nyub.sensor";
+static const int MAX_DATA_LENGTH_BYTES = 4;
+
+void A01nyubComponent::loop() {
+  uint8_t data;
+  while (this->available() > 0) {
+    if (this->read_byte(&data)) {
+      buffer_ += (char) data;
```
Comment (jesserockz) [inline esphome/components/a01nyub/a01nyub.cpp]:
```suggestion
      buffer_.push_back(data);
```

File: esphome/components/a01nyub/a01nyub.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      this->check_buffer_();
+    }
+  }
+}
+
+void A01nyubComponent::check_buffer_() {
+  if (this->buffer_.length() >= MAX_DATA_LENGTH_BYTES) {
```
Comment (jesserockz) [inline esphome/components/a01nyub/a01nyub.cpp]:
```suggestion
  if (this->buffer_.size() >= MAX_DATA_LENGTH_BYTES) {
```

**Comment by jesserockz** [comment]
I have made some minor changes, please verify it still works with your sensor then it can be merged

**Comment by MrSuicideParrot** [comment]
@raccoonia Could you update your node and verify if these changes do not affect the functionality of the pull request?
I can't do it since my sensor went to production, and I won't be able to update its firmware any time soon.

**Comment by raccoonia** [comment]
@MrSuicideParrot HA updated and when I edit my config I see now that my old uart: section has an error and also I get the following
```
Failed config
uart: [source /config/esphome/waterproof-distance.yaml:33]
  Component a01nyub requires this uart bus to declare a tx_pin
  - rx_pin: 
      number: 25
      mode: 
        input: True
        output: False
        open_drain: False
        pullup: False
        pulldown: False
      inverted: False
      drive_strength: 20.0
    baud_rate: 9600
    rx_buffer_size: 256
    stop_bits: 1
    data_bits: 8
    parity: NONE
```
Its obviously because now the tx pin is mandatory, in your documentation you only had the rx definition required.

Should I add a tx pin definition even if its not connected to the a01nyub?

**Comment by MrSuicideParrot** [comment]
@raccoonia Thanks! The problem was introduced with your lasted changes and not HA. I fixed the issue and TX should not be required.
Could you test it again?

**Comment by raccoonia** [comment]
@MrSuicideParrot Hello! I was not very clear before, sorry. By HA update I was meaning that the external component code had already updated on its own, I did now know how often it happens.
So I tested it with a "dummy" tx definition and it worked and right now I commented the tx definition and I can confirm it is no longer required. 
The sensor seems to be working as intended!
I had an issue with the sensor yesterday becoming stale but I am almost certain it was the wifi signal and/or the esp32 getting too hot. Its outdoors and its awfully hot these days
Screenshot from 2023-08-05 20-49-27

Thank you very much for your work to support this sensor!

**Comment by MrSuicideParrot** [comment]
Thanks again for our help @raccoonia!
Now it's only wait for the merge!

**Review by jesserockz — APPROVED** [review]

**Pull request merged**